airflow/providers/databricks/hooks/databricks.py

```diff
@@ -493,3 +504,120 @@ def __init__(self, token: str) -> None:
     def __call__(self, r: PreparedRequest) -> PreparedRequest:
         r.headers['Authorization'] = 'Bearer ' + self.token
         return r
+
+
+class DatabricksAsyncHook(DatabricksHook):
+    """
+    Async version of the ``DatabricksHook``
+    Implements only necessary methods used further in Databricks Triggers.
+    """
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+
+    async def __aenter__(self):
+        self._session = aiohttp.ClientSession()
```

Comment:
This is really good stuff.

Currently you are using one Hook per trigger. That means if you had 32 concurrent triggers, then they would all have their own client sessions.
I've had a look at the docs for aiohttp, it says:
```
it is suggested you use a single session for the lifetime of your application to benefit from connection pooling.
```
https://docs.aiohttp.org/en/stable/client_reference.html

What do you think?

Reply:
Note that hooks are not necessarily run in the same process, so if you want to share sessions among them, you must move the abstraction to the trigger instead.

I wonder how viable it’d be to refactor the sychronous version (`DatabricksHook`) into a [sans I/O](https://sans-io.readthedocs.io/) base class that can be used by `DatabricksExecutionTrigger`, instead of implementing a new, entirely separate hook for it.

Reply:
Ideally each `Triggerer` process would share one session. A DAG can trigger multiple `DatabricksExecutionTrigger` at the same time. Could these triggers all share the same session?

Reply:
They could, yes, because async triggers are designed to run in the same process, that’s way they are async. If they run in their own process, there’s no point to use asyncio in the first place. But the difficult is who should manage that shared session. I wonder if @andrewgodwin actually though about this preemptively 😛

Reply:
Thank you both so much, that's very good points.

Trying to summarise:
- Ideally all tasks executed by each `trigerrer` should share the same session - however, not sure if it'd be possible without touching the core Airflow `Triggerer` functionality.
- >Note that hooks are not necessarily run in the same process, so if you want to share sessions among them, you must move the abstraction to the trigger instead.
  
  Does it mean to move `ClientSession` initialisation to the trigger, i.e. as a property of `DatabricksExecutionTrigger`?
  My understanding is that when there're multiple `triggerer` processes, each trigger will be executed as a separate async task with its own `ClientSession` instance. So even if we move it to `DatabricksExecutionTrigger`, it'd still create a session for each trigger run.
- Refactoring `Databricks` hook - I agree it'd be the perfect solution to keep everything inside a single class. Also, after [the comment](https://github.com/apache/airflow/pull/19736#discussion_r754267058) by @alexott it's even more motivation to do that to prevent making double work in the future. I'll try to refactor it, to minimize repeatable code and isolating IO operations

Reply:
> > you must move the abstraction to the trigger instead.
> My understanding is that when there're multiple `triggerer` processes, each trigger will be executed as a separate async task with its own `ClientSession` instance. So even if we move it to `DatabricksExecutionTrigger`, it'd still create a session for each trigger run.

Sorry, I think I meant to say trigger**er** instead of trigger 🙂

Reply:
Yes, there is no easy mechanism for pooling connections that are all running under a single triggerer right now, but you can also auto-pool from an async hook implementation by detecting same-thread different-coroutine if you need to. Just makes it quite a bit more complex.

Reply:
@andrewgodwin Another issue I encountered with deferrable operators is that task duration during deferral is not counted.
When I run this operator, it says all my DAG's task finish in 1 sec.